runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -848,19 +856,38 @@ void closeBuffer() {
       if (!openBuffer) {
         emit(tag, value);
       } else {
-        bufferState.add(KV.of(tagsToIds.get(tag), value));
+        buffer(KV.of(tagsToIds.get(tag), value));
+      }
+    }
+
+    private void buffer(KV<Integer, WindowedValue<?>> taggedValue) {
```

Comment:
An alternative to pulling the buffer lock dependency in here could be moving the buffer method into DoFnOperator and override that in `ExecutableStageDoFnOperator`.

Reply:
I've pulled out retrieving the lock instead because that's the only dependency we need.